File: payola/web/public/javascripts/cz/payola/web/client/graph/views/GraphView.js
goog.provide('cz.payola.web.client.graph.views.GraphView');
goog.require('cz.payola.web.client.Point');
goog.require('cz.payola.web.client.graph.Constants');
goog.require('cz.payola.web.client.graph.views.EdgeView');
goog.require('cz.payola.web.client.graph.views.VertexView');
goog.require('scala.collection.immutable.List');
goog.require('scala.collection.immutable.Nil');
cz.payola.web.client.graph.views.GraphView = function(graphModel, vertices, edges) {
var self = this;
self.graphModel = graphModel;
self.vertices = vertices;
self.edges = edges;
self.selectedVertexCount = 0;
};
cz.payola.web.client.graph.views.GraphView.prototype.init = function() {
var self = this;
self.vertices = self.createVertices(self.graphModel);
self.edges = self.createEdges(self.graphModel, self.vertices);
};
cz.payola.web.client.graph.views.GraphView.prototype.createVertices = function(graphModel) {
var self = this;
var graphWidth = 200;
var verticesView = scala.collection.immutable.Nil;
var counter = 0;
graphModel.vertices.foreach(function(vertexModel) { var point = new cz.payola.web.client.Point(((counter * 10) % graphWidth), (counter / graphWidth));
var vertexView = new cz.payola.web.client.graph.views.VertexView(counter, point, vertexModel.uri);
verticesView = var $x$1 = verticesView;
scala.collection.immutable.List.$apply(vertexView).$colon$colon$colon($x$1);
;
counter = (counter + 1);
 });
return verticesView;
};
cz.payola.web.client.graph.views.GraphView.prototype.createEdges = function(graphModel, verticesView) {
var self = this;
var edgesView = scala.collection.immutable.Nil;
graphModel.edges.foreach(function(edgeModel) { edgesView = var $x$2 = edgesView;
scala.collection.immutable.List.$apply(new cz.payola.web.client.graph.views.EdgeView(self.lookup(edgeModel.from.uri), self.lookup(edgeModel.to.uri), edgeModel.uri)).$colon$colon$colon($x$2);
;
 });
return edgesView;
};
cz.payola.web.client.graph.views.GraphView.prototype.lookup = function(uri) {
var self = this;
var result = self.vertices.head();
self.vertices.foreach(function(vertex) { (function() {
if ((vertex.information.text == uri)) {
result = vertex;
} else {
}})();
 });
return result;
};
cz.payola.web.client.graph.views.GraphView.prototype.setVertexSelection = function(vertex, selected) {
var self = this;
return (function() {
if ((vertex.selected != selected)) {
self.selectedVertexCount = (self.selectedVertexCount + (function() {
if (selected) {
return 1;
} else {
return -1;
}})());
vertex.selected = selected;
return true;
} else {
return false;
}})();
};
cz.payola.web.client.graph.views.GraphView.prototype.selectVertex = function(vertex) {
var self = this;
return self.setVertexSelection(vertex, true);
};
cz.payola.web.client.graph.views.GraphView.prototype.deselectVertex = function(vertex) {
var self = this;
return self.setVertexSelection(vertex, false);
};
cz.payola.web.client.graph.views.GraphView.prototype.invertVertexSelection = function(vertex) {
var self = this;
return self.setVertexSelection(vertex, (! vertex.selected));
};
cz.payola.web.client.graph.views.GraphView.prototype.deselectAll = function(graph) {
var self = this;
var somethingChanged = false;
(function() {
if ((self.selectedVertexCount > 0)) {
self.vertices.foreach(function(vertex) { somethingChanged = (self.deselectVertex(vertex) || somethingChanged);
 });
} else {
}})();
return somethingChanged;
};
cz.payola.web.client.graph.views.GraphView.prototype.containsSelectedVertex = function() {
var self = this;
return (self.selectedVertexCount > 0);
};
cz.payola.web.client.graph.views.GraphView.prototype.getTouchedVertex = function(p) {
var self = this;
return self.vertices.find(function(v) { return self.isPointInRect(p, v.position.$plus(cz.payola.web.client.graph.Constants.VertexSize.$div(-2.0)), v.position.$plus(cz.payola.web.client.graph.Constants.VertexSize.$div(2.0)));
 });
};
cz.payola.web.client.graph.views.GraphView.prototype.isPointInRect = function(p, topLeft, bottomRight) {
var self = this;
return (p.$greater$eq(topLeft) && p.$less$eq(bottomRight));
};
cz.payola.web.client.graph.views.GraphView.prototype.metaClass_ = new s2js.MetaClass('cz.payola.web.client.graph.views.GraphView', []);

File: payola/web/client/src/main/scala/graph/views/GraphView.scala
package cz.payola.web.client.graph.views

import cz.payola.web.client.graph.Constants._
import cz.payola.web.client.Point
import cz.payola.web.client.graph.model.Graph

class GraphView(val graphModel: Graph, var vertices: List[VertexView], var edges: List[EdgeView])
{
    private var selectedVertexCount = 0

    def init() {
        vertices = createVertices(graphModel)
        edges = createEdges(graphModel, vertices)
    }

    /**
      * Constructs a list of vertexViews based on the graphModel parameter.
      */
    def createVertices(graphModel: Graph): List[VertexView] = {
        val graphWidth = 200 //TODO size of the viewed space

        var verticesView = List[VertexView](/*new VertexView(0, Point.Zero, "plk")*/) //TODO add mutable list to s2js
        var counter = 0
        graphModel.vertices.foreach{ vertexModel =>
            val point = Point((counter*10) % graphWidth, counter/graphWidth)
            val vertexView = new VertexView(counter, point, vertexModel.uri)

            verticesView = verticesView ::: List[VertexView](vertexView)
            counter += 1
        }

        verticesView
    }

    /**
      * Constructs a list of edgeViews based on the graphModel and verticesView parameters.
      */
    def createEdges(graphModel: Graph, verticesView: List[VertexView]): List[EdgeView] = {
        var edgesView = List[EdgeView](/*new EdgeView(null, null, "plk")*/) //TODO add mutable list to s2js
        graphModel.edges.foreach{ edgeModel =>

            edgesView = edgesView ::: List[EdgeView](new EdgeView(
                lookup(edgeModel.from.uri), lookup(edgeModel.to.uri), edgeModel.uri))
        }

        edgesView
    }
    /**
      * DANGEROUS!!! requires the vertices container to be already constructed
      * Searches in the vertices container and returns a vertex with first appearance of
      * the uri in its information.text value.
      */
    def lookup(uri: String): VertexView = {
        //TODO remove when init() is improved
        var result = vertices.head
        vertices.foreach{ vertex =>
            if(vertex.information.text == uri) {
                result = vertex
            }
        }
        result
    }

    def setVertexSelection(vertex: VertexView, selected: Boolean): Boolean = {
        if (vertex.selected != selected) {
            selectedVertexCount += (if (selected) 1 else -1)
            vertex.selected = selected
            true
        } else {
            false
        }
    }

    def selectVertex(vertex: VertexView): Boolean = {
        setVertexSelection(vertex, true)
    }

    def deselectVertex(vertex: VertexView): Boolean = {
        setVertexSelection(vertex, false)
    }

    def invertVertexSelection(vertex: VertexView): Boolean = {
        setVertexSelection(vertex, !vertex.selected)
    }

    def deselectAll(graph: GraphView): Boolean = {
        var somethingChanged = false
        if (selectedVertexCount > 0) {
            vertices.foreach{ vertex =>
                somethingChanged = deselectVertex(vertex) || somethingChanged
            }
        }
        somethingChanged
    }

    def containsSelectedVertex: Boolean = {
        selectedVertexCount > 0
    }

    def getTouchedVertex(p: Point): Option[VertexView] = {
        vertices.find(v => isPointInRect(p, v.position + (VertexSize / -2), v.position + (VertexSize / 2)))
    }

    def isPointInRect(p: Point, topLeft: Point, bottomRight: Point): Boolean = {
        p >= topLeft && p <= bottomRight
    }
}
